### ISSUE TITLE
Multiple codepages in the same file

### ISSUE THREAD

**Issue by gergelyts** [AbsaOSS/cobrix#631]
## Background [Optional]
Hi, 

in our case one file contains several codepages. Different lines are encoded with different code pages. There is a column that acts like a country_code, and it is based on the country code that we know which codepage to use. A file is usually coded with 4-5 codepages. Currently, we would have to read the file for each codepage separately by reading only the appropriate lines (based on the country_code column) with the appropriate codepage. 

This is quite slow. 
## Question
Do you see any possibility to use multiple codepages in the driver at the same time with one readout? 

Thanks!

**Comment by yruslan** [comment]
Currently this is not supported, and I not sure at this point how to add this support in a general way.

But there is a possible workaround. You can make the text field binary by specifying 'USAGE COMP' to that field. It will be converted to `binary` Spark format. And then, you can use Spark, with a UDF to convert the binary field to a string based on the country code.

**Comment by gergelyts** [comment]
Do you have an example code for this? Thanks

**Comment by gergelyts** [comment]
Hi, your workaround does not work, I got the following error:

: scala.MatchError: AlphaNumeric(X(4026),4026,None,Some(EBCDIC),Some(X(4026))) (of class za.co.absa.cobrix.cobol.parser.ast.datatype.AlphaNumeric)

**Comment by yruslan** [comment]
Hi,

you can find an example here: https://github.com/AbsaOSS/cobrix/blob/7462ecf01b6a8ba35250bc4db44a44849643d2cf/spark-cobol/src/test/scala/za/co/absa/cobrix/spark/cobol/source/integration/Test31CobrixExtensionsSpec.scala#L201

Please, use the latest version of Cobrix (2.6.8) as this feature was introduced only recently.

**Comment by Beno922** [comment]
No, out of 10 columns 7 are done by default encoding, and the remaining 3 are decided by country code (1 and 2 bytes codepage). 

In the source application the country code mappings look like this: ([ebcidic_codepage_mapping (1).txt](https://github.com/AbsaOSS/cobrix/files/13165555/ebcidic_codepage_mapping.1.txt)
)

For the third question see also the prior reply.

(These prior requests are related to this [topic:)](url)
https://github.com/AbsaOSS/cobrix/issues/574
https://github.com/AbsaOSS/cobrix/issues/539

If you need more information regarding the source application, feel free to contact @BenceBenedek

**Comment by yruslan** [comment]
Yes, I see. Thank you for the context!

So ideally, you would like a mapping like this:
```json
[
  { 
    "code_field" : "country_code", 
    "target_fields": [ "field1", "field2", "field3" ] 
    "code_mapping": {
      "kor": "ebcdic300",
      "chn": "cp1388"
    }
  }
]
```
so that the encoding of `field1`, `field2` and `field3` is determined by the column `country_code`. When `country_code=kor` use `cp1388`, right?

If multiple country code fields are defined for each record, it can be split like this:
```json
[
  { 
    "code_field" : "country_code1", 
    "target_fields": [ "field1", "field2" ] 
    "code_mapping": {
      "kor": "ebcdic300",
      "chn": "cp1388"
    }
  },
  { 
    "code_field" : "country_code2", 
    "target_fields": [ "field3" ] 
    "code_mapping": {
      "korea": "ebcdic300",
      "china": "cp1388"
    }
  },
]
```

So ideally, if you want to be able to pass such a mapping to Cobrix, and it should figure things out, right?

**Comment by yruslan** [comment]
Now that I think about it, a workaround is possible even now, but not too effective.
```scala
   val df1 = spark.read.format("cobol")
     .option("ebcdic_code_page", "cp037")
     .option("field_code_page:ebcdic300" -> "field1")
     .load("/path/to/files")
     .filter(col("country_code") === "jpn")

   val df2 = spark.read.format("cobol")
     .option("ebcdic_code_page", "cp037")
     .option("field_code_page: cp1388" -> "field1")
     .load("/path/to/files")
     .filter(col("country_code") === "chn")

  val df = df1.union(df2)
```
